Leverage doc value skip lists in DocValuesRewriteMethod if indexed

gsmiller opened the pull request:
### Description

`DocValuesRewriteMethod` (used under-the-hood in `MultiTermQuery`) is another spot where we can leverage the new doc value skip lists when present to potentially skip over doc blocks that don't contain one of the query terms.

Commit:
Leverage doc value skip lists in DocValuesRewriteMethod if indexed

rmuir commented on lucene/core/src/java/org/apache/lucene/document/DocValuesRangeIterator.java:
```diff
@@ -24,8 +24,10 @@
+ * @lucene.experimental
  */
-final class DocValuesRangeIterator extends TwoPhaseIterator {
+public final class DocValuesRangeIterator extends TwoPhaseIterator {
```
move to .search package?

rmuir commented on lucene/core/src/test/org/apache/lucene/search/TestDocValuesRewriteMethod.java:
```diff
@@ -61,14 +61,19 @@ public void setUp() throws Exception {
+          doc.add(new SortedSetDocValuesField(fieldName, new BytesRef(s)));
+        } else {
+          doc.add(SortedSetDocValuesField.indexedField(fieldName, new BytesRef(s)));
+        }
```
can we swap order so that there's no negated comparison. it is always easier to read

gsmiller commented:
It just occurred to me we can’t use DocValuesRangeIterator directly here as it could produce false positive matches. I’ll write a test to capture this and iterate on the PR.

Commit:
cover case where ranges can have gaps

Commit:
add some tests

Commit:
move DocValuesRangeIterator from document -> search pkg

Commit:
spotless

Commit:
add more testing to TestDocValuesRangeIterator

Commit:
changes entry

Commit:
fix missing override declarations

gsmiller commented:
Cleaned up this PR a bit and added testing. Also addressed Robert's feedback (thanks @rmuir). Should be ready for another review if anyone is interested. Thanks!

jpountz commented on lucene/core/src/test/org/apache/lucene/search/TestDocValuesRewriteMethod.java:
```diff
@@ -61,14 +61,19 @@ public void setUp() throws Exception {
                 .setMaxBufferedDocs(TestUtil.nextInt(random(), 50, 1000)));
     List<String> terms = new ArrayList<>();
     int num = atLeast(200);
+    boolean createDocValueSkiplist = random().nextBoolean();
```
Maybe split the test so that we test both cases explicitly every time?

jpountz reviewed (approved):
Nice to see more usage of doc values skippers!

Commit:
expand test coverage in TestDocValuesRewriteMethod

Commit:
oops... fixup unused test variable

Commit:
import order fixup after rebase
